pkg/collector/corechecks/ebpf/c/runtime/oom-kill-kern.c

```diff
@@ -37,39 +38,44 @@ int BPF_KPROBE(kprobe__oom_kill_process, struct oom_control *oc) {
         return 0;
     }
 
-    s->pid = pid;
-    get_cgroup_name(s->cgroup_name, sizeof(s->cgroup_name));
+    // for kernel before 4.11 the prototype for bpf_probe_read helpers
+    // expected a pointer to stack memory. Therefore, we work on stack
+    // variable and update the map value at the end
+    bpf_memcpy(&new, s, sizeof(struct oom_stats));
+
+    new.pid = pid;
+    get_cgroup_name(new.cgroup_name, sizeof(new.cgroup_name));
 
     struct task_struct *p = (struct task_struct *)BPF_CORE_READ(oc, chosen);
     if (!p) {
         return 0;
     }
-    BPF_CORE_READ_INTO(&s->tpid, p, pid);
+    BPF_CORE_READ_INTO(&new.tpid, p, pid);
 
     if (bpf_helper_exists(BPF_FUNC_get_current_comm)) {
-        bpf_get_current_comm(&s->fcomm, sizeof(s->fcomm));
-    }
-    if (bpf_helper_exists(BPF_FUNC_probe_read_str)) {
-        BPF_CORE_READ_STR_INTO(&s->tcomm, p, comm);
-    } else {
-        BPF_CORE_READ_INTO(&s->tcomm, p, comm);
-        s->tcomm[TASK_COMM_LEN - 1] = 0;
+        bpf_get_current_comm(new.fcomm, sizeof(new.fcomm));
     }
 
+    BPF_CORE_READ_INTO(&new.tcomm, p, comm);
+    new.tcomm[TASK_COMM_LEN - 1] = 0;
+
     struct mem_cgroup *memcg = NULL;
 #ifdef COMPILE_CORE
     if (bpf_core_field_exists(oc->totalpages)) {
-        BPF_CORE_READ_INTO(&s->pages, oc, totalpages);
+        BPF_CORE_READ_INTO(&new.pages, oc, totalpages);
     }
     if (bpf_core_field_exists(oc->memcg)) {
         BPF_CORE_READ_INTO(&memcg, oc, memcg);
     }
 #else
-    bpf_probe_read_kernel(&s->pages, sizeof(s->pages), &oc->totalpages);
+    bpf_probe_read_kernel(&new.pages, sizeof(new.pages), &oc->totalpages);
     bpf_probe_read_kernel(&memcg, sizeof(memcg), &oc->memcg);
 #endif
 
-    s->memcg_oom = memcg != NULL ? 1 : 0;
+    new.memcg_oom = memcg != NULL ? 1 : 0;
```

Comment:
looking at this code, I'm not sure why we copy the entire `mem_cgroup` structure when we just need a null check. Could we replace this with `oc->memcg != NULL ? 1 : 0`?

Reply:
We are only reading the pointer to the mem_cgroup structure. We need to do that because `oc->memcg` would result in an invalid memory access since `oc` is kernel memory.

Reply:
Sorry, I meant `&oc->memcg != NULL ? 1 : 0`

Reply:
That is not the test though right?
`&oc->memcg` would be `addrof(oc) + offsetof(oc, memcg)` which will always be non-NULL if oc is non-NULL.
We want to see if the oc->memcg slot holds a non-NULL value.